exceptions/part-2/post.md

```diff
@@ -0,0 +1,157 @@
+
+Tous les développeurs connaissent l'élément de langage de programmation `(throw) exception`, mais les auteurs ci-dessus insistent sur la notion d'erreur : toutes les erreurs ne font pas l'objet d'exceptions. Pour clarifier ce point, je propose d'utiliser les mêmes termes que pour les cas d'utilisation en UML : nominal, alternatif, exceptionnel. Tout le monde est à peu près d'accord sur ce qu'est un cas nominal, c'est quand tout va bien. Par contre, ce qui distingue un cas alternatif et un cas exceptionnel est trop souvent une affaire de goût personnel. Crockford dit que, dans le cas d'un programme qui lit un fichier, le fait que le fichier n'existe pas n'est pas une exception.
+
+Si les développeurs arrivent à se mettre d'accord, de préférence avec l'avis du métier, sur ce qui est une exception et sur ce qui ne l'est pas, le code aura plus de chance d'être lisible. Si l'on suit Michael Feathers, ce sont les cas nominaux et alternatifs qui constituent le corps du programme (`the main logic`). Si une grande partie du code est consacré à la gestion d'erreur, ce ne sont pas des erreurs, mais des cas alternatifs : ils doivent être réintégrés au code principal.
+
+## La règle et ses exceptions : deux exemples
+
+Ces deux cas pratiques concernent les cas alternatifs :
+- l'un utilise les exceptions pour les cas alternatifs dans les use-case (DomainError) en utilisant les exceptions ;
+- l'autre retourne une valeur de retour lors des appels HTTP externes.
+
+### Cas alternatif use-cases - Lever une exception
+
+Lors du traitement d'une requête API, un chemin nominal est attendu :
+- l'utilisateur a les droits d'effectuer cet appel ;
+- les données fournies par l'utilisateur sont valides ;
+- les données existantes sont compatibles avec la demande effectuée par l'utilisateur.
+
+Le corps principal du traitement effectue ces vérifications, mais si l'une d'elle échoue, il initie une exception.
+Prenons l'exemple de l'ajout d'un administrateur, sur un POST de la route `/api/admin/admin-members`. 
+Plusieurs demandes peuvent être envoyées par l'IHM d'administration : connexion instable, plusieurs utilisateurs effectuant la demande en même temps. Ce n'est pas un scénario nominal, mais rien d'exceptionnel non plus. On signale simplement à l'utilisateur que sa demande ne peut aboutir, avec une réponse UnprocessableEntityError (422).
+
+Pourtant, on lève une exception dans le use-case
+```javascript
+throw new AlreadyExistingAdminMemberError();
+```
+[source](https://github.com/1024pix/pix/blob/dev/api/lib/domain/usecases/save-admin-member)
+
+```javascript
+class AlreadyExistingAdminMemberError extends Domain  Error {
+  constructor(message = 'Cet agent a déjà accès') {
+    super(message);
+  }
+}
+```
+[source](https://github.com/1024pix/pix/blob/dev/api/lib/domain/errors.js#L18-L18)
```

Comment:
Globalement pour les liens Github : 
Si nous supprimons les fichiers en question les liens ne fonctionneront plus, il faudrait plutôt utiliser [des permalink ](https://docs.github.com/en/repositories/working-with-files/using-files/getting-permanent-links-to-files),